src/metrics.hpp

```diff
@@ -0,0 +1,428 @@
+/*
+  Copyright (c) 2014-2015 DataStax
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+// Based on implemenations of metrics (especially Meter) from Java library
+// com.codehale.Metrics (https://github.com/dropwizard/metrics)
+
+#ifndef __CASS_METRICS_HPP_INCLUDED__
+#define __CASS_METRICS_HPP_INCLUDED__
+
+#include "scoped_ptr.hpp"
+
+extern "C" {
+#include "third_party/hdr_histogram/hdr_histogram.h"
+}
+
+#include <boost/atomic.hpp>
+
+#include <uv.h>
+
+#include <limits>
+#include <math.h>
+
+namespace cass {
+
+class Metrics {
+private:
+  class ThreadState {
+  public:
+    ThreadState(size_t max_threads)
+      : max_threads_(max_threads)
+      , thread_count_(1) {
+      uv_key_create(&thread_id_key_);
+    }
+
+    ~ThreadState() {
+      uv_key_delete(&thread_id_key_);
+    }
+
+    size_t max_threads() const {
+      return max_threads_;
+    }
+
+    size_t current_thread_id() {
+      void* id = uv_key_get(&thread_id_key_);
+      if (id == NULL) {
+        size_t thread_id = thread_count_++;
+        assert(thread_id <= max_threads_);
+        id = reinterpret_cast<void*>(thread_id);
+        uv_key_set(&thread_id_key_, id);
+      }
+      return reinterpret_cast<size_t>(id) - 1;
+    }
+
+  private:
+    const size_t max_threads_;
+    boost::atomic<size_t> thread_count_;
+    uv_key_t thread_id_key_;
```

Comment:
Are we abandoning libuv 0.10 on this version? This appears to be new in 1.4.2, 0.11.11.